Dockerfile

```diff
@@ -1,21 +1,86 @@
-FROM node:carbon
+# https://blog.hasura.io/an-exhaustive-guide-to-writing-dockerfiles-for-node-js-web-apps-bbee6bd2f3c4
+# https://codefresh.io/docker-tutorial/node_docker_multistage/
 
-RUN mkdir -p /parse-server
-COPY ./ /parse-server/
 
-RUN mkdir -p /parse-server/config
-VOLUME /parse-server/config
+# ------- Stage 1 - Base ---------
+FROM node:carbon-alpine as base
+# apk - https://www.cyberciti.biz/faq/10-alpine-linux-apk-command-examples/
+# RUN apk update && apk upgrade && \
+#     apk add --no-cache git bash
+# python make and g++  - add this for bcrypt 3.x 
 
-RUN mkdir -p /parse-server/cloud
-VOLUME /parse-server/cloud
+# ENV, WORKDIR & COPY commands run as USER root
+# to change ownership run chown (e.g. chown USER node /parse_server) after every command
 
+# after this refer to workdir using ./
 WORKDIR /parse-server
 
-RUN npm install && \
-    npm run build
+# specify multiple volumes in one line - reuse layers during build
+VOLUME ["parse-server/config", "parse-server/cloud"]
 
-ENV PORT=1337
+# copy all package.json-related files
+COPY package.json ./
 
-EXPOSE $PORT
 
-ENTRYPOINT ["npm", "start", "--"]
+# ------- Stage 2 - Dependencies ---------
+# base image for release stage with only prod dependencies
+FROM base AS dependencies
+# set npm configs 
+RUN npm set progress=false && npm config set depth 0
+# install production packages only
+RUN npm install --production 
+
+
+# ------- Stage 3 - Build ---------
+FROM dependencies AS build
+# install all npm required for build (and testing) - saves build time since prod dependencies are already installed
+RUN npm install 
+# copy all context into WORKDIR (/parse-server) excluding items in .dockerignore
+COPY . .
+# Need to run build explicitly as npm will not auto run scripts as ROOT
+# https://stackoverflow.com/questions/47748075/npm-postinstall-not-running-in-docker
+# https://docs.npmjs.com/misc/scripts#user
+RUN npm run prepare && npm run postinstall
+# list all dir/files  - for debugging purposes
+# RUN ls -al
+
+#  ---- UNIT TESTS AND FLOW LINT in stage 3------
+# do not run flow through npm - https://github.com/facebook/flow/issues/3649
+# RUN apk add --no-cache --repository https://nl.alpinelinux.org/alpine/edge/testing flow
+# run unit tests after COPY - any file change will invalidated cached layer and guarantees a new test run
+# make sure flow is removed from pretest in package.json 
+# RUN npm test
+
+
+# ------- Stage 4 - Release ---------
+FROM dependencies AS release
+# lib is the output from babel in the build step
+COPY --from=build /parse-server/lib ./lib
+# copy required files listed in package.json -> files
+COPY /bin ./bin
+COPY /public_html ./public_html
+COPY /views ./views
+# COPY postinstall.js ./
+# COPY PATENTS LICENSE *.md ./
+# list all dir/files  - for debugging purposes
+# RUN ls -al
+
+# capture git_commit in label
+ARG GIT_COMMIT
```

Comment:
is this argument passed by anyone?

Reply:
https://medium.com/microscaling-systems/labelling-automated-builds-on-docker-hub-f3d073fb8e1

I just learned about dockerhub build hooks - https://docs.docker.com/docker-hub/builds/advanced/

I will add a hook to set the arg during build